packages/tangle/tipmanager.go

```diff
@@ -324,9 +327,207 @@ func (t *TipManager) Tips(p payload.Payload, countParents int) (parents MessageI
+
+	for markerWalker.HasNext() && timestampValid {
+		marker := markerWalker.Next().(markers.Marker)
+		timestampValid = t.checkMarker(&marker, messageWalker, markerWalker, minSupportedTimestamp)
+	}
+
+	for messageWalker.HasNext() && timestampValid {
+		timestampValid = t.checkMessage(messageWalker.Next().(MessageID), messageWalker, minSupportedTimestamp)
+	}
+	return timestampValid
+}
+
+func (t *TipManager) processMessage(messageID MessageID, messageWalker, markerWalker *walker.Walker) {
+	t.tangle.Storage.MessageMetadata(messageID).Consume(func(messageMetadata *MessageMetadata) {
+		if messageMetadata.StructureDetails() == nil || messageMetadata.StructureDetails().PastMarkers.Size() == 0 {
+			// need to walk messages
+			messageWalker.Push(messageID)
+			return
+		}
+		messageMetadata.StructureDetails().PastMarkers.ForEach(func(sequenceID markers.SequenceID, index markers.Index) bool {
+			if sequenceID == 0 && index == 0 {
+				// need to walk messages
+				messageWalker.Push(messageID)
+				return false
+			}
+			pastMarker := markers.NewMarker(sequenceID, index)
+			markerWalker.Push(*pastMarker)
+			return true
+		})
+	})
+}
+
+func (t *TipManager) checkMarker(marker *markers.Marker, messageWalker, markerWalker *walker.Walker, minSupportedTimestamp time.Time) (timestampValid bool) {
+	message := t.getMarkerMessage(marker)
+
+	if message.IssuingTime().Before(minSupportedTimestamp) {
+		// marker before minSupportedTimestamp
+		if !t.tangle.ConfirmationOracle.IsMarkerConfirmed(marker) {
+			// if unconfirmed, then incorrect
+			markerWalker.StopWalk()
+			return false
+		}
+
+		// if closest past marker is confirmed and before minSupportedTimestamp, then message should be ok
+		return true
+	}
+	// confirmed after minSupportedTimestamp
+	if t.tangle.ConfirmationOracle.IsMarkerConfirmed(marker) {
+		return true
+	}
+
+	// unconfirmed after minSupportedTimestamp
+
+	// check oldest unconfirmed marker time without walking marker DAG
+	oldestUnconfirmedMarker := t.getOldestUnconfirmedMarker(marker)
+
+	if timestampValid = t.processMarker(marker, minSupportedTimestamp, oldestUnconfirmedMarker); !timestampValid {
+		return
+	}
+
+	t.tangle.Booker.MarkersManager.Manager.Sequence(marker.SequenceID()).Consume(func(sequence *markers.Sequence) {
+		// If there is a confirmed marker before the oldest unconfirmed marker, and it's older than minSupportedTimestamp, need to walk message past cone of oldestUnconfirmedMarker.
+		if sequence.LowestIndex() < oldestUnconfirmedMarker.Index() {
+			confirmedMarkerIdx := t.getPreviousConfirmedIndex(sequence, oldestUnconfirmedMarker.Index()-1)
+
+			confirmedMarkerMsg := t.getMarkerMessage(markers.NewMarker(sequence.ID(), confirmedMarkerIdx))
+			if t.tangle.ConfirmationOracle.IsMessageConfirmed(confirmedMarkerMsg.ID()) && confirmedMarkerMsg.IssuingTime().Before(minSupportedTimestamp) {
+				messageWalker.Push(t.tangle.Booker.MarkersManager.MessageID(oldestUnconfirmedMarker))
+			}
```

Comment:
This is almost the same as L444-448. Maybe we can extract it into a small, descriptive method?